File: VRSampleApp/VRSampleApp/Classes/PresentationLayer/Modules/CustomVideoModule/View/Transition/CustomVideoTransitionManager.m
//
//  CustomVideoTransitionManager.m
//  VRSampleApp
//
//  Copyright © 2017 Ooyala Inc. All rights reserved.
//

#import "CustomVideoTransitionManager.h"
#import "CustomVideoViewController.h"


@interface CustomVideoTransitionManager ()

@property (nonatomic) BOOL presenting;

@end


@implementation CustomVideoTransitionManager

#pragma mark - Private functions

- (void)offStageAnimationViewController:(UIViewController *)animationViewController {
  CustomVideoViewController *customVideoViewController = (CustomVideoViewController *)animationViewController;
  
  customVideoViewController.view.alpha = 0;
  customVideoViewController.containerView.alpha = 0;
  customVideoViewController.containerView.transform = CGAffineTransformMakeTranslation(0, UIScreen.mainScreen.bounds.size.height);
}

- (void)onStageAnimationViewController:(UIViewController *)animationViewController {
  CustomVideoViewController *customVideoViewController = (CustomVideoViewController *)animationViewController;
  
  customVideoViewController.view.alpha = 1;
  customVideoViewController.containerView.alpha = 1;
  customVideoViewController.containerView.transform = CGAffineTransformIdentity;
}

#pragma mark - UIViewControllerAnimatedTransitioning

- (void)animateTransition:(id<UIViewControllerContextTransitioning>)transitionContext {
  UIView *container = transitionContext.containerView;
  
  
  UIViewController *vcFrom = [transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey];
  UIViewController *vcTo = [transitionContext viewControllerForKey:UITransitionContextToViewControllerKey];

  UIViewController *animationViewController;
  UIViewController *bottomViewController;
  
  if (self.presenting) {
    animationViewController = vcTo;
    bottomViewController    = vcFrom;
  } else {
    animationViewController = vcFrom;
    bottomViewController    = vcTo;
  }
  
  UIView *animationView = animationViewController.view;
  UIView *bottomView = animationViewController.view;

  if (self.presenting) {
    [self offStageAnimationViewController:animationViewController];
  }
  
  [container addSubview:bottomView];
  [container addSubview:animationView];
  
  [UIView animateWithDuration:[self transitionDuration:transitionContext]
                        delay:0
                      options:UIViewAnimationOptionCurveEaseInOut animations:^{
                        if (self.presenting) {
                          [self onStageAnimationViewController:animationViewController];
                        } else {
                          [self offStageAnimationViewController:animationViewController];
                        }
                      } completion:^(BOOL finished) {
                        [UIApplication.sharedApplication.keyWindow addSubview:vcTo.view];
                        
                        [transitionContext completeTransition:YES];
                      }];
}

- (NSTimeInterval)transitionDuration:(id<UIViewControllerContextTransitioning>)transitionContext {
  return 0.33;
}

#pragma mark - UIViewControllerTransitioningDelegate

- (nullable id <UIViewControllerAnimatedTransitioning>)animationControllerForPresentedController:(UIViewController *)presented presentingController:(UIViewController *)presenting sourceController:(UIViewController *)source {
  self.presenting = YES;
  
  return self;
}

- (nullable id <UIViewControllerAnimatedTransitioning>)animationControllerForDismissedController:(UIViewController *)dismissed {
  self.presenting = NO;
  
  return self;
}


@end

File: SwiftVRSkinSampleApp/SwiftVRSkinSampleApp/Classes/PresentationLayer/Modules/CustomVideoModule/View/Transition/CustomVideoTransitionManager.swift
//
//  CustomVideoTransitionManager.swift
//  SwiftVRSkinSampleApp
//
//  Copyright © 2017 Ooyala, Inc. All rights reserved.
//

import UIKit

class CustomVideoTransitionManager: NSObject {
  // MARK: - Private properties
  fileprivate var presenting = false

  // MARK: - Private functions
  func offStageAnimationViewController(_ animationViewController: UIViewController) {
    if let animationViewController = animationViewController as? CustomVideoViewController {
      animationViewController.view.alpha              = 0
      animationViewController.containerView.alpha     = 0
      animationViewController.containerView.transform = CGAffineTransform(translationX: 0,
                                                                          y: UIScreen.main.bounds.height)
    }
  }

  func onStageAnimationViewController(_ animationViewController: UIViewController) {
    if let animationViewController = animationViewController as? CustomVideoViewController {
      animationViewController.view.alpha              = 1
      animationViewController.containerView.alpha     = 1
      animationViewController.containerView.transform = .identity
    }
  }
}

// MARK: - UIViewControllerAnimatedTransitioning
extension CustomVideoTransitionManager: UIViewControllerAnimatedTransitioning {

  func animateTransition(using transitionContext: UIViewControllerContextTransitioning) {
    let container = transitionContext.containerView

    let screensTuple: (from: UIViewController, to: UIViewController) =
      (transitionContext.viewController(forKey: UITransitionContextViewControllerKey.from)!,
       transitionContext.viewController(forKey: UITransitionContextViewControllerKey.to)!)

    var animationViewController: UIViewController!
    var bottomViewController: UIViewController!

    if presenting {
      animationViewController = screensTuple.to
      bottomViewController    = screensTuple.from
    } else {
      animationViewController = screensTuple.from
      bottomViewController    = screensTuple.to
    }

    let animationView = animationViewController.view
    let bottomView    = bottomViewController.view

    if presenting {
      offStageAnimationViewController(animationViewController)
    }

    container.addSubview(bottomView!)
    container.addSubview(animationView!)

    UIView.animate(withDuration: transitionDuration(using: transitionContext),
                   delay: 0,
                   options: UIView.AnimationOptions.curveEaseInOut,
                   animations: {
        if self.presenting {
          self.onStageAnimationViewController(animationViewController)
        } else {
          self.offStageAnimationViewController(animationViewController)
        }

    }, completion: { _ in
      UIApplication.shared.keyWindow?.addSubview(screensTuple.to.view)
      transitionContext.completeTransition(true)
    })
  }

  func transitionDuration(using transitionContext: UIViewControllerContextTransitioning?) -> TimeInterval {
    return 0.33
  }
}

// MARK: - UIViewControllerTransitioningDelegate
extension CustomVideoTransitionManager: UIViewControllerTransitioningDelegate {
  func animationController(forPresented presented: UIViewController,
                           presenting: UIViewController,
                           source: UIViewController) -> UIViewControllerAnimatedTransitioning? {
    self.presenting = true
    return self
  }

  func animationController(forDismissed dismissed: UIViewController) -> UIViewControllerAnimatedTransitioning? {
    self.presenting = false
    return self
  }
}
